### PR TITLE
Image hover fix for inverted ranges

### PR DESCRIPTION
**Pull request by jlstevens**
Simple PR to fix the ``ImageView._lrtb`` method, ensuring it returns the correct bounds when the ranges are inverted. Addresses issue #7959.

Here is an example where the y range starts at ten and ends at zero (see the screenshot in #7959 as an example of the incorrect behavior):

image

The code is relatively straightforward although I am not sure I've used the best approach for detecting the inverted ranges.

### CODE REVIEW

**Comment by bryevdv** [comment]
@jlstevens a very little more work will improve this I think. There is an `is_reversed` property on `Range1d`. I'd propose adding that as a required interface to `Range` and then:

* for categorical ranges it just always returns `False`
* for data range it returns the value of `this.flipped`

Then:

* it's easy to add some simple unit tests for these 
* the code in this PR can be simplified to:
  ```
  const xr = this.renderer.xscale.source_range
  const x1 = this._x[i]
  const x2 = xr.is_reversed ? x1 - this._dw[i] : x1 + this._dw[i]
  ```
* this spot can also be fixed up:
  ```
  ./models/plots/plot_canvas.ts:
    467 :       const is_reversed = (rng.start > rng.end)
    485 :         if (is_reversed) {
  ```

**Comment by bryevdv** [comment]
@jlstevens I think that line noted above in `models/plots/plot_canvas.ts` can be fixed up now to use `is_reversed` as well?

File: bokehjs/src/lib/models/glyphs/image.ts [hunk 1]
```diff
@@ -73,10 +73,15 @@ export class ImageView extends XYGlyphView {
...
-    const r = l + this._dw[i]
-    const b = this._y[i]
-    const t = b + this._dh[i]
+    const xr = this.renderer.xscale.source_range
+    const x1 = this._x[i]
+    const x2 = xr.is_reversed ? x1 - this._dw[i] : x1 + this._dw[i]
+
+    const yr = this.renderer.yscale.source_range
+    const y1 = this._y[i]
+    const y2 = yr.is_reversed ? y1 - this._dh[i] : y1 + this._dh[i]
```
Comment (bryevdv) [inline bokehjs/src/lib/models/glyphs/image.ts]: style nit: would say either add a blank line after these three "y" lines of code, or don't add one after the three "x" lines of code

Comment (jlstevens) [reply bokehjs/src/lib/models/glyphs/image.ts]: Addressed in 9fd3fd32d

**Comment by jlstevens** [comment]
@bryevdv I've made the suggested changes including the simplification in ``plot_canvas.ts``.

**Comment by bryevdv** [comment]
Looks great!

**Comment by bryevdv** [comment]
Well, actually, do you mind adding a few tiny unit tests for `is_reversed` ?

**Comment by jlstevens** [comment]
I've added some unit tests for ``is_reversed`` though I just want to check with you what you expect the behavior to be in this case:

```coffeescript
 describe "explicit inverted bounds=(20,10) creation", ->
   r = new DataRange1d({start: 20, end:10})

   it "should be reversed", ->
     expect(r.is_reversed).to.be.equal true
```

This unit test fails because the ``flipped`` property is false even though the start value is greater than the end. Maybe the correct thing to do is to compare the start and end values *after* flipped is potentially applied instead of just returning ``this.flipped``?

**Comment by bryevdv** [comment]
@jlstevens `flipped` is something a user configures, to say "I want auto-range to put end < start" I guess the situation you see in the failed test can arise (and only arise) in the case that a user *initialized* a `DataRange1d` that way (because if the initial start/end values are None, then auto-ranging happens thereafter). 

I guess this simplest and most correct thing to do is put a single `start > end` test on the `Range` and nothing on the subclasses.

**Comment by jlstevens** [comment]
> I guess this simplest and most correct thing to do is put a single start > end test on the Range and nothing on the subclasses.

I agree. It is good to keep the semantics as simple as possible.

I've implemented this suggestion in d8d8e1b and 94f2ccc.

File: bokehjs/test/models/ranges/data_range1d.coffee [hunk 2]
```diff
@@ -67,6 +67,12 @@ describe "datarange1d module", ->
     it "should have max = 20", ->
       expect(r.max).to.be.equal 20
 
+ describe "explicit inverted bounds=(20,10) creation", ->
+   r = new DataRange1d({start: 20, end:10})
+
+   it "should be reversed", ->
+     expect(r.is_reversed).to.be.equal true
+
```
Comment (bryevdv) [inline bokehjs/test/models/ranges/data_range1d.coffee]: Just noting here that ideally we'd also have a test that initializes with `flipped=True` and lets `start` and `end` be filled in by auto-ranging, but that would be a pain to do unfortunately

**Comment by psyrocloud** [comment]
I'm using bokeh 1.0.4 in anaconda, when inverse the y-axis and show a up-down flipped image, the hover tool is showing @image value incorrectly. The value of current pixel is the y-mirror position's pixel value. The example on top of this page cannot test out my problem, because the value of image pixel is same along y axis.

**Comment by thoth291** [comment]
I second the comment above.
```python
from matplotlib import pyplot as plt
from skimage.data import astronaut
c = astronaut().mean(axis=-1)[::2,:]
plt.imshow(c)
plt.show()
```
Gives me this
Screen Shot 2019-03-21 at 10 52 08 PM

While this bokeh (1.0.4) code
```python
scale = 1.0
p = figure(x_range=(1, c.shape[1]), y_range=(c.shape[0],1),
           plot_width=int(c.shape[1]*scale), plot_height=int(c.shape[0]*scale),
           tooltips=[("x", "$x"), ("y", "$y"), ("value", "@image")],
           tools=["hover"])
print(c[89:91,399:401])
p.image(image=[np.flipud(c)],
        x=1, y=c.shape[0],
        dw=c.shape[1], dh=c.shape[0])
show(p)
```
gives me this
Screen Shot 2019-03-21 at 10 59 39 PM

As you can see - image value is not correct - it seems to be reading it from different region of the image.
Hope this reproducer is clear enough to help dig into it.
Please note that it crucially important to keep Y axis increasing from top to bottom.

**Comment by bryevdv** [comment]
Well, I guess this was fixed, then somehow broken again @thoth291 at this point please make a new issue with all that information, so we can track and prioritize this.

**Pull request merged**